pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/services/Merger.scala

```diff
@@ -44,20 +50,39 @@ trait Merger extends MergerLogging {
         }
     }
 
+  implicit class MergeResultAccumulation[T](val result: FieldMergeResult[T]) {
+    def redirectSources: State[MergeState, T] = shouldRedirect(true)
+    def retainSources: State[MergeState, T] = shouldRedirect(false)
+
+    // If the state already contains a source, then don't change the existing `redirect` value
+    // Otherwise, add the source with the current value.
+    private def shouldRedirect(redirect: Boolean): State[MergeState, T] =
+      State { prevState =>
+        val nextState = result.sources.foldLeft(prevState) {
+          case (state, source) if state.contains(source) => state
+          case (state, source)                           => state + (source -> redirect)
+        }
+        (nextState, result.data)
+      }
+  }
+
   def merge(works: Seq[Work[Source]]): MergerOutcome =
     getTargetAndSources(works)
       .map {
         case (target, sources) =>
           logIntentions(target, sources)
           val (mergeResultSources, result) = createMergeResult(target, sources)
-            .run(Set.empty)
+            .run(Map.empty)
             .value
+          val redirectedSources = mergeResultSources.collect {
+            case (source, true) => source
+          }
 
-          val remaining = (sources.toSet -- mergeResultSources)
-            .map(_.transition[Merged](false))
-          val redirects = mergeResultSources
+          val remaining = (sources.toSet -- redirectedSources)
+            .map(_.transition[Merged](0))
+          val redirects = redirectedSources
             .map(redirectSourceToTarget(target))
-            .map(_.transition[Merged](false))
+            .map(_.transition[Merged](0))
```

Comment:
Might be worth some keyword arguments to remind the reader what this parameter is.

Reply:
Cannae do that unfortunately (at least, as far as I can tell, I did look) because this is a generic transition method where the keyword arg would be `args`